pynbody/halo/__init__.py

```diff
@@ -56,6 +57,24 @@ def is_subhalo(self, otherhalo):
 
         return self._halo_catalogue.is_subhalo(self._halo_id, otherhalo._halo_id)
 
+    def physical_units(self, distance='kpc', velocity='km s^-1', mass='Msol', persistent=True):
```

Comment:
There is something a little bit weird about this. If I understand correctly, you are intervening when somebody calls `physical_units` on a halo, to ensure that its properties are also put into physical units. However, I am a bit nervous:

* If `h` is a halo catalogue and I ask for `h[1].physical_units()`, then the arrays of  all halos are converted into physical units (this is unavoidable) but only the properties of halo 1?
* If `f` is the underlying snapshot, `f.physical_units()` doesn't convert any of the halo properties, and indeed there is seemingly no way to convert all of the halo properties

I wonder whether a more appropriate thing to do would be for the halo catalogue `h` to have a `physical_units` method which

* by default, calls `f.physical_units()`
* where supported, converts all the halo properties then calls `f.physical_units()`
* can be called by `h[i].physical_units()`

As a final thought, I see a copy/paste of the code to determine `new_unit` which should instead be factored out somewhere.

Reply:
I agree with all your points. However, there is a technical difficulty here which is that the halo properties are computed just-in-time, so I should need to store somewhere the fact that `f.physical_units()` has been stored. Do you have an idea of how to implement this in a clean way?

Reply:
Yes, there is already a mechanism for this. The snapshot has a variable `_autoconvert` which either contains `None` (for no conversion) or the set of units if `physical_units` has been called previously (with `persist=True`). I think the halos should be able to examine that variable at the point of lazy-computation of the halo properties?

Reply:
This should be good now as per 0df689f and 4fa96b1